### ISSUE TITLE
FusionAuth - Handle www-authenticate challenges as needed

### ISSUE THREAD

**Issue by alex-fusionauth** [nextauthjs/next-auth#8745]
### Provider type

FusionAuth

### Environment

  System:
    OS: macOS 13.4.1
    CPU: (12) arm64 Apple M2 Pro
    Memory: 1.70 GB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.6.1 - /opt/homebrew/bin/node
    npm: 9.8.1 - /opt/homebrew/bin/npm
    pnpm: 8.7.6 - /opt/homebrew/bin/pnpm
  Browsers:
    Brave Browser: 117.1.58.129
    Chrome: 117.0.5938.92
    Safari: 16.5.1
    
  "dependencies": {
    "@auth/core": "latest",
    "@auth/sveltekit": "latest"
  },

### Reproduction URL

https://github.com/alex-fusionauth/fusionauth-sveltekit

### Describe the issue

Trying to find why `Auth.js` is throwing `TODO: Handle www-authenticate challenges as needed`. It appears that in the package `oauth4webapi` the method `parseWwwAuthenticateChallenges` is called and brings back `WWW-Authenticate` header from the code grant response. This has scheme:'basic' listed which causes the error.

Origin of issue
```
    let challenges;
    if ((challenges = o.parseWwwAuthenticateChallenges(codeGrantResponse))) {
        for (const challenge of challenges) {
            console.log("challenge", challenge);
        }
        throw new Error("TODO: Handle www-authenticate challenges as needed");
    }
```

Should the user or FusionAuth be handling this somehow?

### How to reproduce

You will need docker.

1. run `docker compose up -d`
1. cd complete-example
1. copy .env.example to .env
1. npm run dev
1. click sing in button
1. sign in using `richard@example.com` with password: `password`

<img width="733" alt="image" src="https://github.com/nextauthjs/next-auth/assets/139158618/2b29dd67-06d2-4715-a0d3-7c5ff4892198">

Chooose provider while signing in
<img width="471" alt="image" src="https://github.com/nextauthjs/next-auth/assets/139158618/e86df745-0e52-4c16-a11a-feac9f656484">

Error thrown:
<img width="500" alt="image" src="https://github.com/nextauthjs/next-auth/assets/139158618/60656e9f-d61d-468a-b7de-b7694ef4c0b1">

```
challenge { scheme: 'basic', parameters: {} }
[auth][error][CallbackRouteError]: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: Error: TODO: Handle www-authenticate challenges as needed
    at handleOAuth (file:///Users/afa/dev/inversoft/fusionauth/fusionauth-quickstart-javascript-sveltekit-web/complete-application/node_modules/@auth/sveltekit/node_modules/@auth/core/lib/oauth/callback.js:69:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Module.callback (file:///Users/afa/dev/inversoft/fusionauth/fusionauth-quickstart-javascript-sveltekit-web/complete-application/node_modules/@auth/sveltekit/node_modules/@auth/core/lib/routes/callback.js:20:41)
    at async AuthInternal (file:///Users/afa/dev/inversoft/fusionauth/fusionauth-quickstart-javascript-sveltekit-web/complete-application/node_modules/@auth/sveltekit/node_modules/@auth/core/lib/index.js:65:38)
    at async Proxy.Auth (file:///Users/afa/dev/inversoft/fusionauth/fusionauth-quickstart-javascript-sveltekit-web/complete-application/node_modules/@auth/sveltekit/node_modules/@auth/core/index.js:100:30)
    at async Module.respond (/Users/afa/dev/inversoft/fusionauth/fusionauth-quickstart-javascript-sveltekit-web/complete-application/node_modules/@sveltejs/kit/src/runtime/server/respond.js:282:20)
    at async file:///Users/afa/dev/inversoft/fusionauth/fusionauth-quickstart-javascript-sveltekit-web/complete-application/node_modules/@sveltejs/kit/src/exports/vite/dev/index.js:510:22
[auth][details]: {
  "provider": "fusionauth"
}
```

### Expected behavior

There should be no error. Login should continue as the AuthToken is good.

**Comment by alex-fusionauth** [comment]
Just a note I am the DevRel at FusionAuth, happy to work through how to make this valid.

**Comment by marckustech** [comment]
I've been dealing with this Discord provider issue for, like, a whole day now. The weird thing is, when I use it, I don't even get an error message or anything. I just get sent to this `http://localhost:3000/api/auth/error?error=` page, and there are no errors in the URL or in my logs.

The crazy part is, when I deploy it on Vercel, it works perfectly fine. This situation seriously sucks!
![image](https://github.com/nextauthjs/next-auth/assets/122128346/4e274e1a-2e55-46b4-86db-2f2d2af2abe9)

**Comment by ThangHuuVu** [comment]
This is a part that we haven't tacked in the core package. If you are willing to contribute a PR, I'm happy to take a look @alex-fusionauth . It would solve the issue not only for FusionAuth but for all auth providers which return WWWAuthenticateChallenge

**Comment by sohnemann** [comment]
Tried to implement Auth.js with SvelteKit and got stuck here also.
Any idea what to look for?

**Comment by hardylr** [comment]
Same here. I tried to implement Auth.js with SvelteKit using Okta and got stuck here also. How do we handle WWWAuthenticateChallenge?

**Comment by dhadrien** [comment]
Hey @ThangHuuVu, I am also having the issue with the Notion provider. 

Asking some noob qs:
- Does this mean the Notion provider will need a global fix to work? Are there workaround to stil being able to use it?
- Could you please give an update/ more info about this? I doubt I'm knowledgeable enough to help with a PR, but can try

**Comment by hardylr** [comment]
I managed to get Auth.js working with SvelteKit and Okta. Seems like you do not need to set a clientSecret when using Okta when setting token_endpoint_auth_method: "none".  Hope this helps.

```
export const handle = SvelteKitAuth({
  providers: [
    Okta({
          clientId: OKTA_CLIENT_ID,
          //clientSecret: OKTA_CLIENT_SECRET, // DO NOT SET!
          //https://github.com/nextauthjs/next-auth/issues/3540#issuecomment-1021807328
          issuer: OKTA_ISSUER,
          checks: ["pkce", "state"],
          client: {
            token_endpoint_auth_method: "none",
          }          
        }) as Provider
  ]  
})
```

**Comment by luis815** [comment]
Hi everyone!

Not sure if this would help anyone, but I resolved a careless mistake from my end for an error like the above (SvelteKit and Discord):

```txt
challenge {
  scheme: 'basic',
  parameters: { error: 'invalid_client', error_description: '' }
}
[auth][error] CallbackRouteError: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: Error: TODO: Handle www-authenticate challenges as needed
```

The "invalid_client" error indicated that client secret was incorrect.

Cheers!

**Comment by UlasCanAk** [comment]
Hey everyone, running in this same exact issue no matter what I do. Is there any solution to this? I tried following this guide: https://fusionauth.io/docs/quickstarts/quickstart-javascript-nextjs-web#authentication and I'm getting 

`
challenge { scheme: 'basic', parameters: {} }
[auth][error][CallbackRouteError]: Read more at https://errors.authjs.dev#callbackrouteerror
[auth][cause]: Error: TODO: Handle www-authenticate challenges as needed
[auth][details]: {
"provider": "fusionauth"
}
`

I also tried setting token_endpoint_auth_method to "none" and "client_secret_basic", none of them worked sadly. 

I also get an "Invalid URL" error when trying to configure the Provider as it's shown in the guide. Instead I have to provide the  authorization and token endpoint manually:

```
authorization: `https://${env.FUSIONAUTH_ISSUER}/oauth2/authorize`,
token: `https://${env.FUSIONAUTH_ISSUER}/oauth2/token`,
```

I'm on 0.20.0 of @auth/core.
      
      

Thank you!

**Comment by alex-fusionauth** [comment]
> Hey everyone, running in this same exact issue no matter what I do. Is there any solution to this? I tried following this guide: https://fusionauth.io/docs/quickstarts/quickstart-javascript-nextjs-web#authentication and I'm getting
> 
> `challenge { scheme: 'basic', parameters: {} } [auth][error][CallbackRouteError]: Read more at https://errors.authjs.dev#callbackrouteerror [auth][cause]: Error: TODO: Handle www-authenticate challenges as needed [auth][details]: { "provider": "fusionauth" }`
> 
> I also tried setting token_endpoint_auth_method to "none" and "client_secret_basic", none of them worked sadly.
> 
> I also get an "Invalid URL" error when trying to configure the Provider as it's shown in the guide. Instead I have to provide the authorization and token endpoint manually:
> 
> ```
> authorization: `https://${env.FUSIONAUTH_ISSUER}/oauth2/authorize`,
> token: `https://${env.FUSIONAUTH_ISSUER}/oauth2/token`,
> ```
> 
> I'm on 5.0.0-beta.2 of next-auth and 0.20.0 of @auth/core.
> 
> Thank you!

@UlasCanAk This looks like it is a config issue within fusionauth, might be better to post on our community forum or if you ping me on our Slack I can help you directly
https://fusionauth.io/community

**Comment by robotdan** [comment]
Thanks @firebotQL  - I can speak on behalf of FusionAuth. 


In section 4.1 of RFC 7235 where it describes that when a `401` is returned this header must be added, it also says the server may respond with this header for other responses. It isn't explicit, but this sounds like it is ok to return this header on a `200` as well. 
   
> A server MAY generate a WWW-Authenticate header field in other response
   messages to indicate that supplying credentials (or different
   credentials) might affect the response.

- https://datatracker.ietf.org/doc/html/rfc7235#section-4.1

My guess is that is why we are always returning it. This statement in the RFC isn't exactly crystal clear if "other response messages" means other response codes than `401` or something else. If I have the correct understanding (I may not) of "other message", would it looks like this header could be ignored unless the status code is also a `401`? 

I will do some additional research, and we can revisit our current behavior to see if there was any historical reason to always write this header, or if we just made an assumption that it was ok to always return it. 

Thank you for adding this work around while we investigate further! Much appreciated.

**Comment by alex-fusionauth** [comment]
I have added a new Issue/PR to update FusionAuth defaults to better represent our needs. Users should be able to override as they see fit. We can back out the `token.conform` once it is merged into the core release.

Issue: https://github.com/nextauthjs/next-auth/issues/10867
PR: https://github.com/nextauthjs/next-auth/issues/10868
